In [1]:
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import talib
import os
import numpy as np
from scipy.signal import dfreqresp


In [2]:
from binance.client import Client
# sys.path.append(os.path.abspath(".."))  # root /PycharmProjects/MMAT
from config.load_env import load_keys

keys = load_keys()
#print("Loaded keys:", keys)
client = Client(keys['api_key'], keys['secret_key'])

### Load Data from Historical 1yr csv file:

In [3]:
def load_data(csv_path):
    try:
        df = pd.read_csv(csv_path, index_col='timestamp', parse_dates=True)
        df = df[['open', 'high', 'low', 'close', 'volume']].copy()
        print(f"Total K-lines loaded: {len(df)}")
        return df
    except FileNotFoundError:
        print(f"CSV file '{csv_path}' not found.")
        return None

### Resample  to 15 min (Can try other different Time horizon):

In [4]:
def resample_to_15min(df):
    df_15min = df.resample('15min').agg({
        'open': 'first',
        'high': 'max',
        'low': 'min',
        'close': 'last',
        'volume': 'sum'
    }).dropna()
    print(f"Resampled to 15min, total K-lines: {len(df_15min)}")
    return df_15min

### Select Candlestick Patterns for futher combinations with other Indicators 

#### (Below are the 15 ones that I selected based on the accuracy rate >33% from testing all Candlestick Patterns based on historical 15min, and grouped into Bullish & Bearish Patterns) :

In [5]:
def calculate_patterns(df):
    """
    Detect selected high-accuracy candlestick patterns for further analysis.

    Selection Criteria:
    - Based on historical 15-minute BTC K-line backtest
    - Only include patterns with accuracy > 33%

    Returns:
    - df with added pattern columns
    - patterns dictionary for reference
    """
    patterns = {
        #  Bullish Patterns
        'Hammer': talib.CDLHAMMER,
        'InvertedHammer': talib.CDLINVERTEDHAMMER,
        'BullishEngulfing': lambda o, h, l, c: np.where(talib.CDLENGULFING(o, h, l, c) == 100, 100, 0),
        'PiercingLine': talib.CDLPIERCING,
        'MorningStar': talib.CDLMORNINGSTAR,
        'DragonflyDoji': talib.CDLDRAGONFLYDOJI,
        'LongLine': talib.CDLLONGLINE,

        #  Bearish Patterns
        'HangingMan': talib.CDLHANGINGMAN,
        'ShootingStar': talib.CDLSHOOTINGSTAR,
        'BearishEngulfing': lambda o, h, l, c: np.where(talib.CDLENGULFING(o, h, l, c) == -100, -100, 0),
        'DarkCloudCover': talib.CDLDARKCLOUDCOVER,
        'EveningDojiStar': talib.CDLEVENINGDOJISTAR,
        'EveningStar': talib.CDLEVENINGSTAR,
        'GravestoneDoji': talib.CDLGRAVESTONEDOJI,
        'ThreeLineStrike': talib.CDL3LINESTRIKE,
    }

    # Apply pattern detection and store results in df
    for name, func in patterns.items():
        df[name] = func(df['open'].values, df['high'].values, df['low'].values, df['close'].values)

    
    for name in patterns.keys():
        count = (df[name].abs() > 0).sum()
        print(f"{name} detected {count} times.")


    bullish_patterns = ['Hammer', 'InvertedHammer', 'BullishEngulfing', 'PiercingLine',
                        'MorningStar', 'DragonflyDoji', 'LongLine', 'ThreeLineStrike']

    bearish_patterns = ['HangingMan', 'ShootingStar', 'BearishEngulfing', 'DarkCloudCover',
                        'EveningDojiStar', 'EveningStar', 'GravestoneDoji']

    return df, patterns, bullish_patterns, bearish_patterns


In [6]:
# Ensure we only use past data for signal generation
# This function assumes all indicator values are based on past prices only

def generate_signals(df, patterns, window=1):
    """
    Generate candlestick pattern signals using only current and historical data (no future info).
    Includes special handling for patterns like GravestoneDoji.
    """

    # Initialize signal and direction columns
    for name in patterns.keys():
        df[f'Signal_{name}'] = 0
        df[f'Direction_{name}'] = 'NONE'

    # Grouped by typical TA-Lib output behavior
    bullish_patterns_strong = ['BullishEngulfing', 'ThreeLineStrike']
    bullish_patterns = ['Hammer', 'InvertedHammer', 'PiercingLine',
                        'MorningStar', 'DragonflyDoji', 'LongLine']
    bearish_patterns_strong = ['BearishEngulfing']
    bearish_patterns = ['HangingMan', 'ShootingStar', 'DarkCloudCover', 'EveningDojiStar',
                        'EveningStar']  # GravestoneDoji excluded here due to its output behavior

    for i in range(1, len(df) - window):
        for name in patterns.keys():
            value = df[name].iloc[i]

            # Strong bullish
            if name in bullish_patterns_strong and value == 100:
                df.loc[df.index[i], f'Signal_{name}'] = 1
                df.loc[df.index[i], f'Direction_{name}'] = 'UP'

            # General bullish
            elif name in bullish_patterns and value > 0:
                df.loc[df.index[i], f'Signal_{name}'] = 1
                df.loc[df.index[i], f'Direction_{name}'] = 'UP'

            # Strong bearish
            elif name in bearish_patterns_strong and value == -100:
                df.loc[df.index[i], f'Signal_{name}'] = -1
                df.loc[df.index[i], f'Direction_{name}'] = 'DOWN'

            # General bearish
            elif name in bearish_patterns and value < 0:
                df.loc[df.index[i], f'Signal_{name}'] = -1
                df.loc[df.index[i], f'Direction_{name}'] = 'DOWN'

            # Special case: GravestoneDoji outputs 100 but is bearish
            elif name == 'GravestoneDoji' and value == 100:
                df.loc[df.index[i], f'Signal_{name}'] = -1
                df.loc[df.index[i], f'Direction_{name}'] = 'DOWN'

    return df




### Import Indicators:
#### (Below are RSI, MA5days, MA20days, Volume, ATR, Mean ATR, MACD --> feel free to adjust the time period or add other indicators):

### Current Candlestick Pattern + Technical Indicator Trigger Logic (Can Adjust the conditions):
A combined signal is only considered valid when a specific candlestick pattern appears and at least one of the following technical indicator conditions is met: MA, RSI, MACD, or ATR.

| Category          | Candlestick Patterns                                                                 | Technical Indicator Conditions (at least one must be satisfied)              |
|------------------|----------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|
| Bullish (Strong)  | BullishEngulfing, ThreeLineStrike                                                      | MA5 > MA20, RSI > 50, MACD > MACD_Signal, ATR > mean_ATR, High Volume         |
| Bullish (Moderate)| Hammer, InvertedHammer, PiercingLine, MorningStar, DragonflyDoji, LongLine            | Same as above                                                                 |
| Bearish (Strong)  | BearishEngulfing                                                                      | MA5 < MA20, RSI < 45, MACD < MACD_Signal, ATR > mean_ATR, High Volume         |
| Bearish (Moderate)| HangingMan, ShootingStar, DarkCloudCover, EveningDojiStar, EveningStar, GravestoneDoji| Same as above                                                                 |



In [7]:
# Based on top candlestick patterns + indicator to generate signal
import numpy as np
import talib

# Add High Volume Indicator
def is_high_volume(df, i, lookback=20, threshold=1.5):
    if i < lookback:
        return False
    avg_vol = df['volume'].iloc[i - lookback:i].mean()
    return df['volume'].iloc[i] > threshold * avg_vol

# top candlestick pattern + indicator to generate signal
def generate_combined_signals(df, patterns):
    # Pattern 
    bullish_patterns_strong = ['BullishEngulfing', 'ThreeLineStrike']
    bullish_patterns = ['Hammer', 'InvertedHammer', 'PiercingLine',
                        'MorningStar', 'DragonflyDoji', 'LongLine']
    bearish_patterns_strong = ['BearishEngulfing']
    bearish_patterns = ['HangingMan', 'ShootingStar', 'DarkCloudCover',
                        'EveningDojiStar', 'EveningStar', 'GravestoneDoji']

    # initial
    df['bullish_combined'] = 0
    df['bearish_combined'] = 0
    df['bullish_direction'] = 'NONE'
    df['bearish_direction'] = 'NONE'
    df['bullish_trigger'] = ''
    df['bearish_trigger'] = ''

    # Get Indicators -> Can adjust 
    df['MA5'] = talib.SMA(df['close'], timeperiod=5)
    df['MA20'] = talib.SMA(df['close'], timeperiod=20)
    df['RSI'] = talib.RSI(df['close'], timeperiod=14)
    df['ATR'] = talib.ATR(df['high'], df['low'], df['close'], timeperiod=14)
    df['mean_ATR'] = df['ATR'].rolling(window=20).mean()
    df['MACD'], df['MACD_Signal'], _ = talib.MACD(df['close'], fastperiod=12, slowperiod=26, signalperiod=9)

    # Signal columns
    for name in patterns.keys():
        if name == 'GravestoneDoji':
            df[f'Signal_{name}'] = df[name].apply(lambda x: -1 if x == 100 else 0)
        elif name in bullish_patterns_strong:
            df[f'Signal_{name}'] = df[name].apply(lambda x: 1 if x == 100 else 0)
        elif name in bearish_patterns_strong:
            df[f'Signal_{name}'] = df[name].apply(lambda x: -1 if x == -100 else 0)
        else:
            df[f'Signal_{name}'] = df[name].apply(lambda x: 1 if x > 0 else -1 if x < 0 else 0)

    # Signal Trigger Condition
    for i in range(1, len(df) - 1):
        volume_condition = is_high_volume(df, i)
        ma_condition_bull = df['MA5'].iloc[i] > df['MA20'].iloc[i]
        ma_condition_bear = df['MA5'].iloc[i] < df['MA20'].iloc[i]
        rsi_condition_bull = df['RSI'].iloc[i] > 50
        rsi_condition_bear = df['RSI'].iloc[i] < 45
        macd_bull = df['MACD'].iloc[i] > df['MACD_Signal'].iloc[i]
        macd_bear = df['MACD'].iloc[i] < df['MACD_Signal'].iloc[i]
        atr_condition = df['ATR'].iloc[i] > df['mean_ATR'].iloc[i]

        # Bullish pattern
        for pattern in bullish_patterns + bullish_patterns_strong:
            if df[f'Signal_{pattern}'].iloc[i] == 1:
                trigger_parts = [pattern]
                if ma_condition_bull:
                    trigger_parts.append('MA5>MA20')
                if rsi_condition_bull:
                    trigger_parts.append('RSI>50')
                if macd_bull:
                    trigger_parts.append('MACD_Bullish')
                if atr_condition:
                    trigger_parts.append('HighVolatility')
                if volume_condition:
                    trigger_parts.append('HighVolume')
                if len(trigger_parts) >= 2:
                    df.loc[df.index[i], 'bullish_combined'] = 1
                    df.loc[df.index[i], 'bullish_direction'] = 'UP'
                    df.loc[df.index[i], 'bullish_trigger'] = ' + '.join(trigger_parts)

        # Bearish pattern
        for pattern in bearish_patterns + bearish_patterns_strong:
            if df[f'Signal_{pattern}'].iloc[i] == -1:
                trigger_parts = [pattern]
                if ma_condition_bear:
                    trigger_parts.append('MA5<MA20')
                if rsi_condition_bear:
                    trigger_parts.append('RSI<45')
                if macd_bear:
                    trigger_parts.append('MACD_Bearish')
                if atr_condition:
                    trigger_parts.append('HighVolatility')
                if volume_condition:
                    trigger_parts.append('HighVolume')
                if len(trigger_parts) >= 2:
                    df.loc[df.index[i], 'bearish_combined'] = -1
                    df.loc[df.index[i], 'bearish_direction'] = 'DOWN'
                    df.loc[df.index[i], 'bearish_trigger'] = ' + '.join(trigger_parts)
       
    print("Bullish Signal Count:", df['bullish_combined'].sum())
    print("Bearish Signal Count:", -df['bearish_combined'].sum())
    print(df[df['bullish_combined'] == 1][['close', 'bullish_trigger']].tail(5))
    print(df[df['bearish_combined'] == -1][['close', 'bearish_trigger']].tail(5))


    return df


### Evaluation: (Currently use Accuracy rates)

In [8]:
def evaluate_patterns(df, patterns, window=1, threshold=0.0005):
    """
    Evaluate the accuracy of each candlestick pattern signal.

    This function measures by calculating the forward return after each signal and comparing it against a defined threshold.

    Parameters:
    
    window : int, default=1
        Holding period in bars/candles to compute future returns (e.g., 1 bar ahead).

    threshold : float, default=0.0005
        Minimum return required for a signal to be considered successful (e.g., 0.05%).

    """
    results = {}

    #  Compute future return
    df['next_close'] = df['close'].shift(-window)
    df['return'] = (df['next_close'] - df['close']) / df['close']

    # Evaluate each raw candlestick pattern signal
    for name in list(patterns.keys()):
        signal_col = f'Signal_{name}'
        signals = df[df[signal_col] != 0]
        total_signals = len(signals)

        if total_signals == 0:
            results[name] = {'accuracy': 0, 'total_signals': 0, 'correct_signals': 0}
            continue

        correct_signals = len(signals[
            ((signals[signal_col] == 1) & (df.loc[signals.index, 'return'] >= threshold)) |
            ((signals[signal_col] == -1) & (df.loc[signals.index, 'return'] <= -threshold))
        ])
        accuracy = correct_signals / total_signals * 100

        results[name] = {
            'accuracy': accuracy,
            'total_signals': total_signals,
            'correct_signals': correct_signals
        }

    #  Evaluate combined signals (pattern + indicator)
    for signal_col in ['bullish_combined', 'bearish_combined']:
        signals = df[df[signal_col] != 0]
        total_signals = len(signals)

        if total_signals == 0:
            results[signal_col] = {'accuracy': 0, 'total_signals': 0, 'correct_signals': 0}
            continue

        correct_signals = len(signals[
            ((signals[signal_col] == 1) & (df.loc[signals.index, 'return'] >= threshold)) |
            ((signals[signal_col] == -1) & (df.loc[signals.index, 'return'] <= -threshold))
        ])
        accuracy = correct_signals / total_signals * 100

        results[signal_col] = {
            'accuracy': accuracy,
            'total_signals': total_signals,
            'correct_signals': correct_signals
        }

    return results


### Plot result:

In [9]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def plot_combined_results(df, symbol, data_range=2000):
    df_plot = df.iloc[-data_range:].copy()

    up_signals = df_plot[df_plot['bullish_combined'] == 1]
    down_signals = df_plot[df_plot['bearish_combined'] == -1]
    print(f"Plotting Combined: {len(up_signals)} Bullish UP signals, {len(down_signals)} Bearish DOWN signals in last {data_range} K-lines")

    fig = make_subplots(
        rows=4, cols=1,
        shared_xaxes=True,
        vertical_spacing=0.1,
        subplot_titles=['Candlestick + MA', 'RSI', 'ATR', 'Volume'],
        row_heights=[0.4, 0.2, 0.2, 0.2]
    )

    # Candlestick + MA
    fig.add_trace(go.Candlestick(
        x=df_plot.index,
        open=df_plot['open'],
        high=df_plot['high'],
        low=df_plot['low'],
        close=df_plot['close'],
        name='Candlestick',
        increasing_line_color='green',
        decreasing_line_color='red'
    ), row=1, col=1)

    fig.add_trace(go.Scatter(
        x=df_plot.index, y=df_plot['MA5'], mode='lines', name='5 MA', line=dict(color='blue')
    ), row=1, col=1)
    fig.add_trace(go.Scatter(
        x=df_plot.index, y=df_plot['MA20'], mode='lines', name='20 MA', line=dict(color='purple')
    ), row=1, col=1)

    fig.add_trace(go.Scatter(
        x=up_signals.index,
        y=up_signals['close'] * 1.005,
        mode='markers',
        marker=dict(symbol='triangle-up', color='green', size=10),
        name='Bullish Combined Signal',
        text=up_signals['bullish_trigger'],
        hoverinfo='text+x+y'
    ), row=1, col=1)

    fig.add_trace(go.Scatter(
        x=down_signals.index,
        y=down_signals['close'] * 0.995,
        mode='markers',
        marker=dict(symbol='triangle-down', color='red', size=10),
        name='Bearish Combined Signal',
        text=down_signals['bearish_trigger'],
        hoverinfo='text+x+y'
    ), row=1, col=1)

    # RSI
    fig.add_trace(go.Scatter(
        x=df_plot.index, y=df_plot['RSI'], mode='lines', name='RSI', line=dict(color='blue')
    ), row=2, col=1)
    fig.add_hline(y=50, line_dash='dash', line_color='black', row=2, col=1)

    # ATR
    fig.add_trace(go.Scatter(
        x=df_plot.index, y=df_plot['ATR'], mode='lines', name='ATR', line=dict(color='orange')
    ), row=3, col=1)
    fig.add_trace(go.Scatter(
        x=df_plot.index, y=df_plot['mean_ATR'] * 1.2, mode='lines', name='1.2 * mean_ATR', line=dict(color='red', dash='dash')
    ), row=3, col=1)

    # Volume
    fig.add_trace(go.Bar(
        x=df_plot.index, y=df_plot['volume'], name='Volume', marker_color='blue'
    ), row=4, col=1)

    # Layout
    fig.update_layout(
        title=f'Historical 15 min Combined Signals for {symbol} with MA, RSI, ATR, and Volume',
        xaxis_title='Time',
        yaxis_title='Price ($)',
        yaxis2_title='RSI',
        yaxis3_title='ATR',
        yaxis4_title='Volume',
        xaxis_rangeslider_visible=False,
        showlegend=True,
        height=800,
        template='plotly_white'
    )

    # Only open in browser
    fig.show(renderer="browser")


### Def Main:

In [10]:
def main():
    #  Load 1-min raw data
    csv_path='/Users/wynn/PycharmProjects/MMAT/notebooks/btc_1min.csv'
    df = load_data(csv_path)
    if df is None:
        return

    # Resample to 15-min data
    df = resample_to_15min(df)

    #  Apply candlestick pattern detection using TA-Lib
    df, patterns, bullish_patterns, bearish_patterns = calculate_patterns(df)

    # Generate basic pattern-only signals (e.g., Signal_Hammer, Signal_Engulfing)
    df = generate_signals(df, patterns)

    # Generate combined signals (pattern + indicator confirmation)
    df = generate_combined_signals(df, patterns)

    #  Evaluate signal accuracy based on forward returns
    accuracy_results = evaluate_patterns(df, patterns)

    # pattern evaluation results
    print("\n--- Pattern Signal Accuracy ---")
    for name, metrics in sorted(accuracy_results.items(), key=lambda x: x[1]['accuracy'], reverse=True):
        print(f"{name} - Accuracy: {metrics['accuracy']:.2f}%, Total: {metrics['total_signals']}, Correct: {metrics['correct_signals']}")

    # Plot signal charts
    #plot_pattern_results(df, patterns, 'BTC')
    plot_combined_results(df, 'BTC')

if __name__ == "__main__":
    main()


Total K-lines loaded: 526000
Resampled to 15min, total K-lines: 35067
Hammer detected 1058 times.
InvertedHammer detected 134 times.
BullishEngulfing detected 1008 times.
PiercingLine detected 3 times.
MorningStar detected 87 times.
DragonflyDoji detected 634 times.
LongLine detected 6718 times.
HangingMan detected 590 times.
ShootingStar detected 161 times.
BearishEngulfing detected 1006 times.
DarkCloudCover detected 1 times.
EveningDojiStar detected 20 times.
EveningStar detected 81 times.
GravestoneDoji detected 546 times.
ThreeLineStrike detected 82 times.
Bullish Signal Count: 5219
Bearish Signal Count: 1775
                        close  \
timestamp                       
2025-04-30 16:30:00  94086.95   
2025-04-30 17:15:00  94036.62   
2025-04-30 17:30:00  94267.14   
2025-04-30 20:00:00  94638.60   
2025-04-30 22:00:00  94584.66   

                                                       bullish_trigger  
timestamp                                                               


### BTC 15-Min Combined Signal Evaluation

#### Combined Signal Summary
> Signals are triggered only when a candlestick pattern occurs AND at least one technical indicator (MA, RSI, MACD, ATR) confirms.

#### Bullish Combined Signals
- **Total Signals**: 5,219

**Recent 5 Bullish Triggers**:
| Timestamp           | Close Price | Trigger Description |
|---------------------|-------------|----------------------|
| 2025-04-30 16:30:00 | 94,086.95   | BullishEngulfing + MACD_Bullish + HighVolatility |
| 2025-04-30 17:15:00 | 94,036.62   | LongLine + MACD_Bullish + HighVolatility |
| 2025-04-30 17:30:00 | 94,267.14   | LongLine + MA5>MA20 + RSI>50 + MACD_Bullish + HighVolatility |
| 2025-04-30 20:00:00 | 94,638.60   | LongLine + MA5>MA20 + RSI>50 + MACD_Bullish + HighVolatility |
| 2025-04-30 22:00:00 | 94,584.66   | BullishEngulfing + MA5>MA20 + RSI>50 + MACD_Bullish |

---

####  Bearish Combined Signals
- **Total Signals**: 1,775

**Recent 5 Bearish Triggers**:
| Timestamp           | Close Price | Trigger Description |
|---------------------|-------------|----------------------|
| 2025-04-29 18:45:00 | 95,391.29   | HangingMan + HighVolatility |
| 2025-04-30 00:15:00 | 94,400.01   | GravestoneDoji + MA5<MA20 |
| 2025-04-30 09:00:00 | 94,705.63   | GravestoneDoji + MA5<MA20 + MACD_Bearish |
| 2025-04-30 13:00:00 | 94,282.01   | BearishEngulfing + MA5<MA20 + RSI<45 + MACD_Bearish |
| 2025-04-30 19:15:00 | 93,797.90   | BearishEngulfing + RSI<45 |

---

####  Combined Signal Accuracy

| Signal Type         | Accuracy | Total Signals | Correct Signals |
|---------------------|----------|----------------|-----------------|
| **bullish_combined**| 37.25%   | 5,219          | 1,944           |
| **bearish_combined**| 39.04%   | 1,775          | 693             |

---
